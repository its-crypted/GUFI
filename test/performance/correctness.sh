#!/usr/bin/env bash
# This file is part of GUFI, which is part of MarFS, which is released
# under the BSD license.
#
#
# Copyright (c) 2017, Los Alamos National Security (LANS), LLC
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation and/or
# other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors
# may be used to endorse or promote products derived from this software without
# specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#
# From Los Alamos National Security, LLC:
# LA-CC-15-039
#
# Copyright (c) 2017, Los Alamos National Security, LLC All rights reserved.
# Copyright 2017. Los Alamos National Security, LLC. This software was produced
# under U.S. Government contract DE-AC52-06NA25396 for Los Alamos National
# Laboratory (LANL), which is operated by Los Alamos National Security, LLC for
# the U.S. Department of Energy. The U.S. Government has rights to use,
# reproduce, and distribute this software.  NEITHER THE GOVERNMENT NOR LOS
# ALAMOS NATIONAL SECURITY, LLC MAKES ANY WARRANTY, EXPRESS OR IMPLIED, OR
# ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.  If software is
# modified to produce derivative works, such modified software should be
# clearly marked, so as not to confuse it with the version available from
# LANL.
#
# THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.



set -e

ROOT="$(realpath ${BASH_SOURCE[0]})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"

DB="test_perf.db"
OUTPUT="correctness.out"

function cleanup() {
    rm -f "${DB}"
}

trap cleanup EXIT

cleanup

(
echo "Initialize database file ${DB}"
${ROOT}/contrib/performance/initialize.py "${DB}" gufi_query > /dev/null
echo

echo "Tables:"
sqlite3 "${DB}" "SELECT '    ' || name FROM sqlite_master WHERE type=='table';" | sort
echo

echo "Views:"
sqlite3 "${DB}" "SELECT '    ' || name FROM sqlite_master WHERE type=='view';"  | sort
echo

echo "Create a configuration"
config_hash=$(${ROOT}/contrib/performance/configuration.py --add "${DB}" gufi_query "testindex.gufi")
echo "    Hash: ${config_hash}"
echo "Number of configurations available:" $(sqlite3 "${DB}" "SELECT COUNT(*) FROM configurations")
echo

echo "Insert raw numbers"
declare -a commits=("commit1" "commit2")
branch="test"

for run in $(seq 1 5)
do
    # only RealTime has values - all columns should be calculated the same way
    sqlite3 "${DB}" "INSERT INTO raw_numbers VALUES ('${config_hash}', '${commits[0]}', '${branch}', ${run}, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ${run}, 0);"

    sqlite3 "${DB}" "INSERT OR IGNORE INTO commits VALUES (NULL, '${commits[0]}');"
done

for run in $(seq 2 6)
do
    # only RealTime has values - all columns should be calculated the same way
    sqlite3 "${DB}" "INSERT INTO raw_numbers VALUES ('${config_hash}', '${commits[1]}', '${branch}', ${run}, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ${run}, 0);"

    sqlite3 "${DB}" "INSERT OR IGNORE INTO commits VALUES (NULL, '${commits[1]}');"
done

# there should be 5 runs for 2 commits = 10 rows of raw numbers
echo "Sets of raw numbers:" $(sqlite3 "${DB}" "SELECT COUNT(*) FROM raw_numbers;")
echo

# 2 rows - one for each commit
echo "Rows in average:" $(sqlite3 "${DB}" "SELECT COUNT(*) FROM average;")
echo "Rows in max:    " $(sqlite3 "${DB}" "SELECT COUNT(*) FROM max;")
echo "Rows in min:    " $(sqlite3 "${DB}" "SELECT COUNT(*) FROM min;")
echo

for commit in "${commits[@]}"
do
    echo "${commit} average:" $(sqlite3 "${DB}" "SELECT RealTime FROM average WHERE git=='${commit}';")
    echo "${commit} max:    " $(sqlite3 "${DB}" "SELECT RealTime FROM max     WHERE git=='${commit}';")
    echo "${commit} min:    " $(sqlite3 "${DB}" "SELECT RealTime FROM min     WHERE git=='${commit}';")
    echo
done

# test comparisons
echo "Compare average"
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" average RealTime=-0.5,0.5
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" average RealTime=-1,1
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" average RealTime=2,3
echo

echo "Compare max"
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" max RealTime=-0.5,0.5
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" max RealTime=-1,1
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" max RealTime=2,3
echo

echo "Compare min"
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" min RealTime=-0.5,0.5
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" min RealTime=-1,1
${ROOT}/contrib/performance/compare.py "${DB}" gufi_query "${commits[0]}" "${commits[1]}" min RealTime=2,3
echo

echo "Dump average"
${ROOT}/contrib/performance/dump.py "${DB}" gufi_query "${config_hash}" average
echo

echo "Dump max"
${ROOT}/contrib/performance/dump.py "${DB}" gufi_query "${config_hash}" max
echo

echo "Dump min"
${ROOT}/contrib/performance/dump.py "${DB}" gufi_query "${config_hash}" min
echo
) 2>&1 | tee "${OUTPUT}"

diff ${ROOT}/test/performance/correctness.expected "${OUTPUT}"
rm "${OUTPUT}"
