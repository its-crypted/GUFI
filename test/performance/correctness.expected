Initialize database file test_perf.db

Tables:
    commits
    configurations
    raw_numbers

Views:
    average
    max
    median
    min
    stddev

Create a configuration
    Hash: f617d2c24f25742941df37b02c24cf755060c906936aef05247d157c661a868e0deafe955be4d6e837bee8e4fa9ff4035bb1cf41c32a1ced18e16dc6cbd7a6d7
Number of configurations available: 1

Insert raw numbers
Sets of raw numbers: 10

Rows in average: 2
Rows in max:     2
Rows in min:     2

commit1 average: 3.0
commit1 max:     5.0
commit1 min:     1.0

commit2 average: 4.0
commit2 max:     6.0
commit2 min:     2.0

Compare average
RealTime 3.0 4.0 increased
RealTime 3.0 4.0 same
RealTime 3.0 4.0 decreased

Compare max
RealTime 5.0 6.0 increased
RealTime 5.0 6.0 same
RealTime 5.0 6.0 decreased

Compare min
RealTime 1.0 2.0 increased
RealTime 1.0 2.0 same
RealTime 1.0 2.0 decreased

Dump average
1 commit1 setup_globals 0.0
1 commit1 setup_aggregate 0.0
1 commit1 work 0.0
1 commit1 opendir 0.0
1 commit1 opendb 0.0
1 commit1 sqlite3_open 0.0
1 commit1 set_pragmas 0.0
1 commit1 load_extension 0.0
1 commit1 addqueryfuncs 0.0
1 commit1 descend 0.0
1 commit1 check_args 0.0
1 commit1 level 0.0
1 commit1 level_branch 0.0
1 commit1 while_branch 0.0
1 commit1 readdir 0.0
1 commit1 readdir_branch 0.0
1 commit1 strncmp 0.0
1 commit1 strncmp_branch 0.0
1 commit1 snprintf 0.0
1 commit1 lstat 0.0
1 commit1 isdir 0.0
1 commit1 isdir_branch 0.0
1 commit1 access 0.0
1 commit1 set_struct 0.0
1 commit1 clone 0.0
1 commit1 pushdir 0.0
1 commit1 attach 0.0
1 commit1 sqltsumcheck 0.0
1 commit1 sqltsum 0.0
1 commit1 sqlsum 0.0
1 commit1 sqlent 0.0
1 commit1 detach 0.0
1 commit1 closedb 0.0
1 commit1 closedir 0.0
1 commit1 utime 0.0
1 commit1 free_work 0.0
1 commit1 output_timestamps 0.0
1 commit1 aggregate 0.0
1 commit1 output 0.0
1 commit1 cleanup_globals 0.0
1 commit1 rows 0.0
1 commit1 query_count 0.0
1 commit1 RealTime 3.0
1 commit1 ThreadTime 0.0
2 commit2 setup_globals 0.0
2 commit2 setup_aggregate 0.0
2 commit2 work 0.0
2 commit2 opendir 0.0
2 commit2 opendb 0.0
2 commit2 sqlite3_open 0.0
2 commit2 set_pragmas 0.0
2 commit2 load_extension 0.0
2 commit2 addqueryfuncs 0.0
2 commit2 descend 0.0
2 commit2 check_args 0.0
2 commit2 level 0.0
2 commit2 level_branch 0.0
2 commit2 while_branch 0.0
2 commit2 readdir 0.0
2 commit2 readdir_branch 0.0
2 commit2 strncmp 0.0
2 commit2 strncmp_branch 0.0
2 commit2 snprintf 0.0
2 commit2 lstat 0.0
2 commit2 isdir 0.0
2 commit2 isdir_branch 0.0
2 commit2 access 0.0
2 commit2 set_struct 0.0
2 commit2 clone 0.0
2 commit2 pushdir 0.0
2 commit2 attach 0.0
2 commit2 sqltsumcheck 0.0
2 commit2 sqltsum 0.0
2 commit2 sqlsum 0.0
2 commit2 sqlent 0.0
2 commit2 detach 0.0
2 commit2 closedb 0.0
2 commit2 closedir 0.0
2 commit2 utime 0.0
2 commit2 free_work 0.0
2 commit2 output_timestamps 0.0
2 commit2 aggregate 0.0
2 commit2 output 0.0
2 commit2 cleanup_globals 0.0
2 commit2 rows 0.0
2 commit2 query_count 0.0
2 commit2 RealTime 4.0
2 commit2 ThreadTime 0.0

Dump max
1 commit1 setup_globals 0.0
1 commit1 setup_aggregate 0.0
1 commit1 work 0.0
1 commit1 opendir 0.0
1 commit1 opendb 0.0
1 commit1 sqlite3_open 0.0
1 commit1 set_pragmas 0.0
1 commit1 load_extension 0.0
1 commit1 addqueryfuncs 0.0
1 commit1 descend 0.0
1 commit1 check_args 0.0
1 commit1 level 0.0
1 commit1 level_branch 0.0
1 commit1 while_branch 0.0
1 commit1 readdir 0.0
1 commit1 readdir_branch 0.0
1 commit1 strncmp 0.0
1 commit1 strncmp_branch 0.0
1 commit1 snprintf 0.0
1 commit1 lstat 0.0
1 commit1 isdir 0.0
1 commit1 isdir_branch 0.0
1 commit1 access 0.0
1 commit1 set_struct 0.0
1 commit1 clone 0.0
1 commit1 pushdir 0.0
1 commit1 attach 0.0
1 commit1 sqltsumcheck 0.0
1 commit1 sqltsum 0.0
1 commit1 sqlsum 0.0
1 commit1 sqlent 0.0
1 commit1 detach 0.0
1 commit1 closedb 0.0
1 commit1 closedir 0.0
1 commit1 utime 0.0
1 commit1 free_work 0.0
1 commit1 output_timestamps 0.0
1 commit1 aggregate 0.0
1 commit1 output 0.0
1 commit1 cleanup_globals 0.0
1 commit1 rows 0
1 commit1 query_count 0
1 commit1 RealTime 5.0
1 commit1 ThreadTime 0.0
2 commit2 setup_globals 0.0
2 commit2 setup_aggregate 0.0
2 commit2 work 0.0
2 commit2 opendir 0.0
2 commit2 opendb 0.0
2 commit2 sqlite3_open 0.0
2 commit2 set_pragmas 0.0
2 commit2 load_extension 0.0
2 commit2 addqueryfuncs 0.0
2 commit2 descend 0.0
2 commit2 check_args 0.0
2 commit2 level 0.0
2 commit2 level_branch 0.0
2 commit2 while_branch 0.0
2 commit2 readdir 0.0
2 commit2 readdir_branch 0.0
2 commit2 strncmp 0.0
2 commit2 strncmp_branch 0.0
2 commit2 snprintf 0.0
2 commit2 lstat 0.0
2 commit2 isdir 0.0
2 commit2 isdir_branch 0.0
2 commit2 access 0.0
2 commit2 set_struct 0.0
2 commit2 clone 0.0
2 commit2 pushdir 0.0
2 commit2 attach 0.0
2 commit2 sqltsumcheck 0.0
2 commit2 sqltsum 0.0
2 commit2 sqlsum 0.0
2 commit2 sqlent 0.0
2 commit2 detach 0.0
2 commit2 closedb 0.0
2 commit2 closedir 0.0
2 commit2 utime 0.0
2 commit2 free_work 0.0
2 commit2 output_timestamps 0.0
2 commit2 aggregate 0.0
2 commit2 output 0.0
2 commit2 cleanup_globals 0.0
2 commit2 rows 0
2 commit2 query_count 0
2 commit2 RealTime 6.0
2 commit2 ThreadTime 0.0

Dump min
1 commit1 setup_globals 0.0
1 commit1 setup_aggregate 0.0
1 commit1 work 0.0
1 commit1 opendir 0.0
1 commit1 opendb 0.0
1 commit1 sqlite3_open 0.0
1 commit1 set_pragmas 0.0
1 commit1 load_extension 0.0
1 commit1 addqueryfuncs 0.0
1 commit1 descend 0.0
1 commit1 check_args 0.0
1 commit1 level 0.0
1 commit1 level_branch 0.0
1 commit1 while_branch 0.0
1 commit1 readdir 0.0
1 commit1 readdir_branch 0.0
1 commit1 strncmp 0.0
1 commit1 strncmp_branch 0.0
1 commit1 snprintf 0.0
1 commit1 lstat 0.0
1 commit1 isdir 0.0
1 commit1 isdir_branch 0.0
1 commit1 access 0.0
1 commit1 set_struct 0.0
1 commit1 clone 0.0
1 commit1 pushdir 0.0
1 commit1 attach 0.0
1 commit1 sqltsumcheck 0.0
1 commit1 sqltsum 0.0
1 commit1 sqlsum 0.0
1 commit1 sqlent 0.0
1 commit1 detach 0.0
1 commit1 closedb 0.0
1 commit1 closedir 0.0
1 commit1 utime 0.0
1 commit1 free_work 0.0
1 commit1 output_timestamps 0.0
1 commit1 aggregate 0.0
1 commit1 output 0.0
1 commit1 cleanup_globals 0.0
1 commit1 rows 0
1 commit1 query_count 0
1 commit1 RealTime 1.0
1 commit1 ThreadTime 0.0
2 commit2 setup_globals 0.0
2 commit2 setup_aggregate 0.0
2 commit2 work 0.0
2 commit2 opendir 0.0
2 commit2 opendb 0.0
2 commit2 sqlite3_open 0.0
2 commit2 set_pragmas 0.0
2 commit2 load_extension 0.0
2 commit2 addqueryfuncs 0.0
2 commit2 descend 0.0
2 commit2 check_args 0.0
2 commit2 level 0.0
2 commit2 level_branch 0.0
2 commit2 while_branch 0.0
2 commit2 readdir 0.0
2 commit2 readdir_branch 0.0
2 commit2 strncmp 0.0
2 commit2 strncmp_branch 0.0
2 commit2 snprintf 0.0
2 commit2 lstat 0.0
2 commit2 isdir 0.0
2 commit2 isdir_branch 0.0
2 commit2 access 0.0
2 commit2 set_struct 0.0
2 commit2 clone 0.0
2 commit2 pushdir 0.0
2 commit2 attach 0.0
2 commit2 sqltsumcheck 0.0
2 commit2 sqltsum 0.0
2 commit2 sqlsum 0.0
2 commit2 sqlent 0.0
2 commit2 detach 0.0
2 commit2 closedb 0.0
2 commit2 closedir 0.0
2 commit2 utime 0.0
2 commit2 free_work 0.0
2 commit2 output_timestamps 0.0
2 commit2 aggregate 0.0
2 commit2 output 0.0
2 commit2 cleanup_globals 0.0
2 commit2 rows 0
2 commit2 query_count 0
2 commit2 RealTime 2.0
2 commit2 ThreadTime 0.0

